delegate-app/HydraAuction/Delegate/Tracing.hs

```diff
@@ -0,0 +1,73 @@
+{-# LANGUAGE FunctionalDependencies #-}
+{-# LANGUAGE RankNTypes #-}
+
+module HydraAuction.Delegate.Tracing (
+  -- * tracing types
+
+  -- ** tracing transformer
+  TracerT (..),
+  runWithTracer,
+  runWithTracer',
+  askTracer,
+
+  -- ** tracing type class
+  MonadTracer (..),
+) where
+
+-- Prelude imports
+import Hydra.Prelude (ReaderT (runReaderT))
+import Prelude
+
+-- Haskell imports
+import Control.Monad.Reader (
+  MonadIO,
+  MonadReader (ask),
+  MonadTrans (lift),
+ )
+import Control.Monad.State (MonadState)
+import Control.Monad.Writer (MonadWriter)
+import Control.Tracer (Tracer, natTracer, traceWith)
+
+{- | a monad that one can trace in
+   the typeclass takes as argument the
+   type that serves as the trace and the monad
+   that you can trace in
+-}
+class MonadTracer t m | m -> t where
```

Comment:
We already have one such interface. So you need integrate this one into existing one. Like make it superclass.

And I am still not understanding why fundeps are better here. Type family have one obvious advantage - it does not require to write second variable. What is advantage of fundeps?

Reply:
The obvious advantage is that you don't need to specify an equality constraint if you want a specific type to trace with.

Reply:
Okay, you are right, that may be actual usecase if we will kill `logMsg`. And we will need ability to make message types non fixed per monad, by embedding one into another. That is also a way to kill `logMsg` - just introduce String message type instance.

Reply:
Separated this to follow-up issue: https://github.com/mlabs-haskell/hydra-auction/issues/231